File: be/Controllers/NotificationController.cs
using FA23_Convocation2023_API.DTO;
using FA23_Convocation2023_API.Models;
using FA23_Convocation2023_API.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using System.Security.Claims;

namespace FA23_Convocation2023_API.Controllers
{
    [Route("api/[controller]")]
    [ApiController]
    [Authorize]
    public class NotificationController : ControllerBase
    {
        private readonly NotificationService _notificationService;
        private readonly Convo24Context _context;

        public NotificationController(NotificationService notificationService, Convo24Context context)
        {
            _notificationService = notificationService;
            _context = context;
        }

        private string GetCurrentUserId()
        {
            return User.FindFirst("userID")?.Value ?? "";
        }

        private string GetCurrentUserRole()
        {
            return User.FindFirst("role")?.Value ?? "";
        }

        // GET: api/Notification
        [HttpGet]
        public async Task<ActionResult<NotificationListResponse>> GetNotifications(
            [FromQuery] int pageIndex = 1,
            [FromQuery] int pageSize = 10,
            [FromQuery] string? status = null)
        {
            try
            {
                Console.WriteLine($"[DEBUG] Getting notifications - pageIndex: {pageIndex}, pageSize: {pageSize}, status: {status}");
                var notifications = await _notificationService.GetAllNotificationsAsync(pageIndex, pageSize, status);
                Console.WriteLine($"[DEBUG] Found {notifications.Count} notifications");
                var response = new List<NotificationResponse>();
                foreach (var n in notifications)
                {
                    // Manually load related entities if needed
                    var hall = n.HallId.HasValue ? await _context.Halls.FindAsync(n.HallId.Value) : null;
                    var session = n.SessionId.HasValue ? await _context.Sessions.FindAsync(n.SessionId.Value) : null;
                    var createdByUser = await _context.Users.FindAsync(n.CreatedBy);
                    var broadcastByUser = !string.IsNullOrEmpty(n.BroadcastBy) ? await _context.Users.FindAsync(n.BroadcastBy) : null;

                    response.Add(new NotificationResponse
                    {
                        NotificationId = n.NotificationId,
                        Title = n.Title,
                        Content = n.Content,
                        Priority = n.Priority,
                        PriorityText = GetPriorityText(n.Priority),
                        HallId = n.HallId,
                        HallName = hall?.HallName,
                        SessionId = n.SessionId,
                        SessionNumber = session?.Session1,
                        CreatedBy = n.CreatedBy,
                        CreatedByName = createdByUser?.FullName ?? "",
                        BroadcastBy = n.BroadcastBy,
                        BroadcastByName = broadcastByUser?.FullName,
                        CreatedAt = n.CreatedAt,
                        ScheduledAt = n.ScheduledAt,
                        BroadcastAt = n.BroadcastAt,
                        Status = n.Status,
                        IsAutomatic = n.IsAutomatic,
                        RepeatCount = n.RepeatCount,
                        Scope = GetNotificationScope(n, hall, session)
                    });
                }

                return Ok(new NotificationListResponse
                {
                    Notifications = response,
                    TotalCount = response.Count, // TODO: Get actual count from service
                    PageIndex = pageIndex,
                    PageSize = pageSize,
                    TotalPages = (int)Math.Ceiling((double)response.Count / pageSize)
                });
            }
            catch (Exception ex)
            {
                return StatusCode(500, new { message = "Error retrieving notifications", error = ex.Message });
            }
        }

        // GET: api/Notification/{id}
        [HttpGet("{id}")]
        public async Task<ActionResult<NotificationResponse>> GetNotification(int id)
        {
            try
            {
                var notification = await _notificationService.GetNotificationByIdAsync(id);
                if (notification == null)
                {
                    return NotFound(new { message = "Notification not found" });
                }

                var response = new NotificationResponse
                {
                    NotificationId = notification.NotificationId,
                    Title = notification.Title,
                    Content = notification.Content,
                    Priority = notification.Priority,
                    PriorityText = GetPriorityText(notification.Priority),
                    HallId = notification.HallId,
                    HallName = notification.Hall?.HallName,
                    SessionId = notification.SessionId,
                    SessionNumber = notification.Session?.Session1,
                    CreatedBy = notification.CreatedBy,
                    CreatedByName = notification.CreatedByUser?.FullName ?? "",
                    BroadcastBy = notification.BroadcastBy,
                    BroadcastByName = notification.BroadcastByUser?.FullName,
                    CreatedAt = notification.CreatedAt,
                    ScheduledAt = notification.ScheduledAt,
                    BroadcastAt = notification.BroadcastAt,
                    Status = notification.Status,
                    IsAutomatic = notification.IsAutomatic,
                    RepeatCount = notification.RepeatCount,
                    Scope = GetNotificationScope(notification)
                };

                return Ok(response);
            }
            catch (Exception ex)
            {
                return StatusCode(500, new { message = "Error retrieving notification", error = ex.Message });
            }
        }

        // POST: api/Notification
        [HttpPost]
        [Authorize(Roles = "MN,NO")] // Manager or Noticer can create
        public async Task<ActionResult<NotificationResponse>> CreateNotification([FromBody] CreateNotificationRequest request)
        {
            try
            {
                var notification = new Notification
                {
                    Title = request.Title,
                    Content = request.Content,
                    Priority = request.Priority,
                    HallId = request.HallId,
                    SessionId = request.SessionId,
                    ScheduledAt = request.ScheduledAt,
                    IsAutomatic = request.IsAutomatic,
                    RepeatCount = request.RepeatCount,
                    CreatedBy = GetCurrentUserId()
                };

                var createdNotification = await _notificationService.CreateNotificationAsync(notification);

                return CreatedAtAction(
                    nameof(GetNotification),
                    new { id = createdNotification.NotificationId },
                    new { message = "Notification created successfully", notificationId = createdNotification.NotificationId }
                );
            }
            catch (Exception ex)
            {
                return StatusCode(500, new { message = "Error creating notification", error = ex.Message });
            }
        }

        // PUT: api/Notification/{id}
        [HttpPut("{id}")]
        [Authorize(Roles = "MN,NO")]
        public async Task<ActionResult> UpdateNotification(int id, [FromBody] UpdateNotificationRequest request)
        {
            try
            {
                var updatedNotification = new Notification
                {
                    Title = request.Title,
                    Content = request.Content,
                    Priority = request.Priority,
                    HallId = request.HallId,
                    SessionId = request.SessionId,
                    ScheduledAt = request.ScheduledAt,
                    IsAutomatic = request.IsAutomatic,
                    RepeatCount = request.RepeatCount
                };

                var result = await _notificationService.UpdateNotificationAsync(id, updatedNotification);
                if (result == null)
                {
                    return NotFound(new { message = "Notification not found" });
                }

                return Ok(new { message = "Notification updated successfully" });
            }
            catch (Exception ex)
            {
                return StatusCode(500, new { message = "Error updating notification", error = ex.Message });
            }
        }

        // DELETE: api/Notification/{id}
        [HttpDelete("{id}")]
        [Authorize(Roles = "MN")]
        public async Task<ActionResult> DeleteNotification(int id)
        {
            try
            {
                var result = await _notificationService.DeleteNotificationAsync(id);
                if (!result)
                {
                    return NotFound(new { message = "Notification not found" });
                }

                return Ok(new { message = "Notification deleted successfully" });
            }
            catch (Exception ex)
            {
                return StatusCode(500, new { message = "Error deleting notification", error = ex.Message });
            }
        }

        // POST: api/Notification/{id}/broadcast
        [HttpPost("{id}/broadcast")]
        [Authorize(Roles = "NO")] // Only Noticer can broadcast
        public async Task<ActionResult> StartBroadcast(int id)
        {
            try
            {
                var result = await _notificationService.StartBroadcastAsync(id, GetCurrentUserId());
                if (!result)
                {
                    return BadRequest(new { message = "Cannot start broadcast. Notification not found or not in pending status." });
                }

                return Ok(new { message = "Broadcast started successfully" });
            }
            catch (Exception ex)
            {
                return StatusCode(500, new { message = "Error starting broadcast", error = ex.Message });
            }
        }

        // POST: api/Notification/{id}/complete
        [HttpPost("{id}/complete")]
        [Authorize(Roles = "NO")]
        public async Task<ActionResult> CompleteBroadcast(int id)
        {
            try
            {
                var result = await _notificationService.CompleteBroadcastAsync(id);
                if (!result)
                {
                    return BadRequest(new { message = "Cannot complete broadcast. Notification not found or not broadcasting." });
                }

                return Ok(new { message = "Broadcast completed successfully" });
            }
            catch (Exception ex)
            {
                return StatusCode(500, new { message = "Error completing broadcast", error = ex.Message });
            }
        }

        // POST: api/Notification/{id}/cancel
        [HttpPost("{id}/cancel")]
        [Authorize(Roles = "MN,NO")]
        public async Task<ActionResult> CancelNotification(int id)
        {
            try
            {
                var result = await _notificationService.CancelNotificationAsync(id);
                if (!result)
                {
                    return BadRequest(new { message = "Cannot cancel notification. Notification not found or already completed." });
                }

                return Ok(new { message = "Notification cancelled successfully" });
            }
            catch (Exception ex)
            {
                return StatusCode(500, new { message = "Error cancelling notification", error = ex.Message });
            }
        }

        // GET: api/Notification/pending
        [HttpGet("pending")]
        [Authorize(Roles = "NO")]
        public async Task<ActionResult<List<NotificationResponse>>> GetPendingNotifications(
            [FromQuery] int? hallId = null,
            [FromQuery] int? sessionId = null)
        {
            try
            {
                var notifications = await _notificationService.GetNotificationsByHallAndSessionAsync(hallId, sessionId);
                var response = notifications.Select(n => new NotificationResponse
                {
                    NotificationId = n.NotificationId,
                    Title = n.Title,
                    Content = n.Content,
                    Priority = n.Priority,
                    PriorityText = GetPriorityText(n.Priority),
                    HallId = n.HallId,
                    HallName = n.Hall?.HallName,
                    SessionId = n.SessionId,
                    SessionNumber = n.Session?.Session1,
                    CreatedAt = n.CreatedAt,
                    ScheduledAt = n.ScheduledAt,
                    Status = n.Status,
                    IsAutomatic = n.IsAutomatic,
                    RepeatCount = n.RepeatCount,
                    Scope = GetNotificationScope(n)
                }).ToList();

                return Ok(response);
            }
            catch (Exception ex)
            {
                return StatusCode(500, new { message = "Error retrieving pending notifications", error = ex.Message });
            }
        }

        private static string GetPriorityText(int priority)
        {
            return priority switch
            {
                1 => "High",
                2 => "Medium",
                3 => "Low",
                _ => "Medium"
            };
        }

        private static string GetNotificationScope(Notification notification, Hall? hall = null, Session? session = null)
        {
            if (notification.HallId == null && notification.SessionId == null)
                return "Toàn trường";

            if (notification.HallId != null && notification.SessionId != null)
                return $"Hall: {hall?.HallName}, Session: {session?.Session1}";

            if (notification.HallId != null)
                return $"Hall: {hall?.HallName}";

            if (notification.SessionId != null)
                return $"Session: {session?.Session1}";

            return "Toàn trường";
        }
    }
}

File: be-nest/src/notification/notification.controller.ts
import {
  BadRequestException,
  Body,
  Controller,
  Delete,
  Get,
  NotFoundException,
  Param,
  ParseIntPipe,
  Post,
  Put,
  Query,
} from '@nestjs/common';
import { ApiTags } from '@nestjs/swagger';
import { Type } from 'class-transformer';
import { IsInt, IsOptional, IsString, Min } from 'class-validator';
import { Permission } from '../auth/permissions.js';
import type { ActorContext } from '../common/guards/actor-context.js';
import { CurrentActor } from '../common/guards/current-actor.decorator.js';
import { RequirePermissions } from '../common/guards/require-permissions.decorator.js';
import { RequireRoles } from '../common/guards/require-roles.decorator.js';
import { NotificationDto } from './dto/notification.dto.js';
import { NotificationService } from './notification.service.js';

class NotificationListQuery {
  @Type(() => Number)
  @IsInt()
  @Min(1)
  pageIndex = 1;

  @Type(() => Number)
  @IsInt()
  @Min(1)
  pageSize = 10;

  @IsOptional()
  @IsString()
  status?: string;
}

class PendingNotificationQuery {
  @Type(() => Number)
  @IsOptional()
  @IsInt()
  hallId?: number;

  @Type(() => Number)
  @IsOptional()
  @IsInt()
  sessionId?: number;
}

@ApiTags('notification')
@Controller('Notification')
export class NotificationController {
  constructor(private readonly notifications: NotificationService) {}

  @Get()
  list(@Query() query: NotificationListQuery) {
    return this.notifications.list(query.pageIndex, query.pageSize, query.status);
  }

  @Get('pending')
  @RequireRoles('NO')
  pending(@Query() query: PendingNotificationQuery) {
    return this.notifications.pending(query.hallId, query.sessionId);
  }

  @Get(':id')
  async get(@Param('id', ParseIntPipe) id: number) {
    const notification = await this.notifications.get(id);
    if (!notification) {
      throw new NotFoundException({
        code: 'notification/not-found',
        message: 'Không tìm thấy thông báo.',
      });
    }
    return notification;
  }

  @Post()
  @RequirePermissions(Permission.ManageNotifications)
  async create(@Body() input: NotificationDto, @CurrentActor() actor: ActorContext) {
    const created = await this.notifications.create(input, actor.userId);
    return { message: 'Tạo thông báo thành công.', notificationId: created.id };
  }

  @Put(':id')
  @RequirePermissions(Permission.ManageNotifications)
  async update(@Param('id', ParseIntPipe) id: number, @Body() input: NotificationDto) {
    if (!(await this.notifications.update(id, input))) {
      throw new NotFoundException({
        code: 'notification/not-found',
        message: 'Không tìm thấy thông báo.',
      });
    }
    return { message: 'Cập nhật thông báo thành công.' };
  }

  @Delete(':id')
  @RequirePermissions(Permission.ManageNotifications)
  async delete(@Param('id', ParseIntPipe) id: number) {
    if (!(await this.notifications.delete(id))) {
      throw new NotFoundException({
        code: 'notification/not-found',
        message: 'Không tìm thấy thông báo.',
      });
    }
    return { message: 'Xóa thông báo thành công.' };
  }

  @Post(':id/start-broadcast')
  @RequirePermissions(Permission.BroadcastNotifications)
  async startBroadcast(
    @Param('id', ParseIntPipe) id: number,
    @CurrentActor() actor: ActorContext,
  ) {
    if (!(await this.notifications.transition(id, ['PENDING'], 'BROADCASTING', actor.userId))) {
      throw new BadRequestException({
        code: 'notification/invalid-transition',
        message: 'Không thể bắt đầu phát: thông báo không tồn tại hoặc không ở trạng thái chờ.',
      });
    }
    return { message: 'Bắt đầu phát thông báo thành công.' };
  }

  @Post(':id/complete')
  @RequirePermissions(Permission.BroadcastNotifications)
  async complete(@Param('id', ParseIntPipe) id: number) {
    if (!(await this.notifications.transition(id, ['BROADCASTING'], 'COMPLETED'))) {
      throw new BadRequestException({
        code: 'notification/invalid-transition',
        message: 'Không thể hoàn tất: thông báo không tồn tại hoặc chưa được phát.',
      });
    }
    return { message: 'Hoàn tất phát thông báo thành công.' };
  }

  @Post(':id/cancel')
  @RequirePermissions(Permission.ManageNotifications)
  async cancel(@Param('id', ParseIntPipe) id: number) {
    if (!(await this.notifications.transition(id, ['PENDING', 'BROADCASTING'], 'CANCELLED'))) {
      throw new BadRequestException({
        code: 'notification/invalid-transition',
        message: 'Không thể hủy: thông báo không tồn tại hoặc đã hoàn tất.',
      });
    }
    return { message: 'Hủy thông báo thành công.' };
  }

  @Post(':id/broadcast')
  @RequirePermissions(Permission.BroadcastNotifications)
  async broadcast(@Param('id', ParseIntPipe) id: number) {
    const data = await this.notifications.broadcast(id);
    if (!data) {
      throw new NotFoundException({
        code: 'notification/not-found',
        message: 'Không tìm thấy thông báo.',
      });
    }
    return { message: 'Phát thông báo đến người đọc thành công.', data };
  }
}
